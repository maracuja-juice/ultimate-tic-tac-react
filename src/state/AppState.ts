import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
