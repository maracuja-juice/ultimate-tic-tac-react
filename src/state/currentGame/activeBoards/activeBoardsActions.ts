import { Point } from '../../../util/Point';

export const SET_ALLOWED_BOARDS = 'activeBoardsReducer/set-allowed-boards';
export const CALCULATE_ALLOWED_BOARDS = 'activeBoardsReducer/calculate-allowed-boards';

export const setAllowedBoards = ( boards: Point[], ) => ({
    type: SET_ALLOWED_BOARDS,
    payload: boards
});
export const calculateActiveBoards = ( lastMove: Point ) => ({
    type: CALCULATE_ALLOWED_BOARDS,
    payload: lastMove
});