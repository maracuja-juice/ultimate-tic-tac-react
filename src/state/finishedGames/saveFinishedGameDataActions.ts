import { FinishedGameState } from '../AppState';

export const SAVE_GAME_DATA = 'save-game-data/save-game-data';
export const SAVE_GAME_DATA_PENDING = 'save-game-data/save-game-data-pending';
export const SAVE_GAME_DATA_FULFILLED = 'save-game-data/save-game-data-fulfilled';
export const SAVE_GAME_DATA_REJECTED = 'save-game-data/save-game-data-rejected';

export const saveGameData = ( finishedGame: FinishedGameState ): SaveGameDataAction => ({
    type: SAVE_GAME_DATA,
    payload: finishedGame
});

export const saveGameDataPending = ( ): SaveGameDataPendingAction => ({
    type: SAVE_GAME_DATA_PENDING
});

export const saveGameDataFulfilled = ( ): SaveGameDataFulfilledAction => ({
    type: SAVE_GAME_DATA_FULFILLED,
});

export const saveGameDataRejected = ( errorMessage: string ): SaveGameDataRejectedAction => ({
    type: SAVE_GAME_DATA_REJECTED,
    payload: errorMessage
});

export interface SaveGameDataAction {
    type: string;
    payload: FinishedGameState;
}

export interface SaveGameDataPendingAction {
    type: string;
}

export interface SaveGameDataFulfilledAction {
    type: string;
}

export interface SaveGameDataRejectedAction {
    type: string;
    payload: string;
}