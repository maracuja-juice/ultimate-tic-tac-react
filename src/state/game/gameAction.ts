import { Player } from '../AppState';
import { Point } from '../../util/Point';

export const CHANGE_PLAYER = 'game-reducer/player-changed';
export const PLAYER_MOVED = 'game/player-moved';
export const PLAYER_MOVED_VALID = 'game/player-moved-valid';
export const GAME_FINISHED = 'game/game-finished';
export const CHECK_GAME_FINISHED = 'game/check-game-finished';
export const RESTART_GAME = 'restart-game';

export const changePlayer = () => ({
    type: CHANGE_PLAYER
});

export const restartGame = () => ({
    type: RESTART_GAME
});

export const gameFinished = ( winningPlayer?: Player | null ) => ({
    type: GAME_FINISHED,
    payload: winningPlayer
});

export const checkGameFinished = (  ) => ({
    type: CHECK_GAME_FINISHED
});

export const playerMoved = ( boardPosition: Point, tilePosition: Point ): PlayerMovedAction => ({
    type: PLAYER_MOVED,
    payload: {
        boardPosition: boardPosition,
        tilePosition: tilePosition
    }
});

export interface PlayerMovedAction {
    type: string;
    payload: {
        boardPosition: Point
        tilePosition: Point
    };
}
