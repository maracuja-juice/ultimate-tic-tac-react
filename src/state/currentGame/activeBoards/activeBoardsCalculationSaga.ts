import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CALCULATE_ALLOWED_BOARDS, setAllowedBoards } from './activeBoardsActions';
import { getBoards } from '../../selectors/AppStateSelectors';
import { getNewActiveBoards } from '../../../util/ActiveBoards';

function* calculateActiveBoards( action: GenericAction ) {
    const boards = yield select( getBoards );
    const lastMove = action.payload;

    const activeBoards = getNewActiveBoards(lastMove, boards);

    yield put( setAllowedBoards( activeBoards ) );
}

function* activeBoardsCalculationSaga() {
    yield takeEvery( CALCULATE_ALLOWED_BOARDS, calculateActiveBoards );
}

export default activeBoardsCalculationSaga;