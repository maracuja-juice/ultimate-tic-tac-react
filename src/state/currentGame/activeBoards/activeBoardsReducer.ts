import { Point, TicTacToeGame } from '../../../lib';
import { GenericAction } from '../../AppState';
import { SET_ACTIVE_BOARDS } from './activeBoardsActions';
import { RESTART_GAME } from '../../commonAction';

const initialState: Point[] = new TicTacToeGame([]).getCurrentActiveBoards();

const activeBoardsReducer = (state = initialState, action: GenericAction) => {
  switch (action.type) {
    case SET_ACTIVE_BOARDS: {
      return action.payload;
    }
    case RESTART_GAME: {
      return initialState;
    }
    default: {
      return state;
    }
  }
};

export default activeBoardsReducer;
