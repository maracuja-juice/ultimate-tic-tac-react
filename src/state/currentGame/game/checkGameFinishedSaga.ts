import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CHECK_GAME_FINISHED, gameFinished } from './gameAction';
import { getMoves } from '../../selectors/AppStateSelectors';
import { setAllowedBoards } from '../activeBoards/activeBoardsActions';
import { saveGameData } from '../../finishedGames/saveFinishedGameDataActions';
import { getFinishedGameData } from '../../selectors/FinishedGameStateSelectors';
import { TicTacToeGame } from '../../../util';

function* checkIfGameFinished(action: GenericAction) {
  const moves = yield select(getMoves);

  const winResult = new TicTacToeGame(moves).getWinResult();
  const isGameFinished = winResult.isFinished;
  const winningPlayer = winResult.winningPlayer;

  if (isGameFinished) {
    yield put(gameFinished(winningPlayer));
    yield put(setAllowedBoards([]));
    yield put(saveGameData(yield select(getFinishedGameData)));
  }
}

function* checkGameFinishedSaga() {
  yield takeEvery(CHECK_GAME_FINISHED, checkIfGameFinished);
}

export default checkGameFinishedSaga;
