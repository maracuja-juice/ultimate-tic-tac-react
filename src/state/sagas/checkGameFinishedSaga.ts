import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction, SmallBoardInformation } from '../AppState';
import { CHECK_GAME_FINISHED, gameFinished } from '../game/gameAction';
import { getBoards } from '../selectors/AppStateSelectors';
import { getWinResult } from '../../util/CheckBoard';

function* checkIfGameFinished( action: GenericAction ) {
    const boards = yield select( getBoards );

    // TODO: make this map more beautiful or change the params that the getWinResult function takes
    const mappedBoards = boards.map( (smallBoard: SmallBoardInformation) => {
        return {smallBoardPoint: smallBoard.point, value: smallBoard.value};
    });
    const winResult = getWinResult(mappedBoards);

    const isGameFinished = winResult.isFinished;
    const winningPlayer = winResult.winningPlayer;

    if (isGameFinished) {
        yield put( gameFinished( winningPlayer! ) );
    }
}

function* checkGameFinishedSaga() {
    yield takeEvery( CHECK_GAME_FINISHED, checkIfGameFinished );
}

export default checkGameFinishedSaga;