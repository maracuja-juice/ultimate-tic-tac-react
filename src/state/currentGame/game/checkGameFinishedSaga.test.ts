import { expectSaga } from 'redux-saga-test-plan';
import checkGameFinishedSaga from './checkGameFinishedSaga';
import { getBoards } from '../../selectors/AppStateSelectors';
import { select } from 'redux-saga/effects';
import { CHECK_GAME_FINISHED, GAME_FINISHED } from './gameAction';
import { Player } from '../../AppState';
import { SET_ALLOWED_BOARDS } from '../activeBoards/activeBoardsActions';
import {
  circleFinishedBoardMock,
  crossFinishedBoardMock,
  unfinishedBoardMock,
  circleFinishedGameMock,
  crossFinishedGameMock,
} from '../../../__mocks__';
import { SAVE_GAME_DATA } from '../../finishedGames/saveFinishedGameDataActions';
import { getFinishedGameData } from '../../selectors/FinishedGameStateSelectors';

describe('checkGameFinishedSaga', () => {
  it(
    'should dispatch gameFinished action with Circle and' +
      ' setAllowedBoards to no board when the game is finished',
    () => {
      return expectSaga(checkGameFinishedSaga)
        .provide([
          [select(getBoards), circleFinishedBoardMock],
          [select(getFinishedGameData), circleFinishedGameMock],
        ])
        .put({ type: GAME_FINISHED, payload: Player.Circle })
        .put({ type: SET_ALLOWED_BOARDS, payload: [] })
        .put({ type: SAVE_GAME_DATA, payload: circleFinishedGameMock })
        .dispatch({ type: CHECK_GAME_FINISHED })
        .silentRun();
    },
  );

  it(
    'should dispatch gameFinished action with Cross and setAllowedBoards to no board ' +
      'if the game is finished',
    () => {
      return expectSaga(checkGameFinishedSaga)
        .provide([
          [select(getBoards), crossFinishedBoardMock],
          [select(getFinishedGameData), crossFinishedGameMock],
        ])
        .put({ type: GAME_FINISHED, payload: Player.Cross })
        .put({ type: SET_ALLOWED_BOARDS, payload: [] })
        .put({ type: SAVE_GAME_DATA, payload: crossFinishedGameMock })
        .dispatch({ type: CHECK_GAME_FINISHED })
        .silentRun();
    },
  );

  it('should dispatch no action if the game is not finished', () => {
    return expectSaga(checkGameFinishedSaga)
      .provide([[select(getBoards), unfinishedBoardMock]])
      .dispatch({ type: CHECK_GAME_FINISHED })
      .silentRun();
  });

  // if more put effects happen: this catches it + this checks for the order
  it('should match snapshot', () => {
    return expectSaga(checkGameFinishedSaga)
      .provide([
        [select(getBoards), circleFinishedBoardMock],
        [select(getFinishedGameData), circleFinishedGameMock],
      ])
      .dispatch({ type: CHECK_GAME_FINISHED })
      .silentRun()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
