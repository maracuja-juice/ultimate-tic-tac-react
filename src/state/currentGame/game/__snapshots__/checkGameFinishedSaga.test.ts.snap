// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkGameFinishedSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "game/check-game-finished",
          "@@redux-saga-test-plan/json/function/checkIfGameFinished",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          Object {
            "type": "game/check-game-finished",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": 1,
          "type": "game/game-finished",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Array [],
          "type": "activeBoardsReducer/set-active-boards",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "gameState": Array [
              Object {
                "position": Object {
                  "x": 0,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 1,
                  },
                ],
                "value": 1,
              },
              Object {
                "position": Object {
                  "x": 0,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 1,
              },
              Object {
                "position": Object {
                  "x": 0,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 1,
                  },
                ],
                "value": 1,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 0,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 1,
                  },
                ],
                "value": 1,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 1,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 0,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 0,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
            ],
            "moves": Array [
              Object {
                "boardPosition": Object {
                  "x": 1,
                  "y": 0,
                },
                "moveNumber": 1,
                "player": 0,
                "tilePosition": Object {
                  "x": 1,
                  "y": 1,
                },
              },
            ],
            "winner": "O",
          },
          "saveOnline": true,
          "type": "save-game-data/save-game-data",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/getMoves",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "game/check-game-finished",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "game/check-game-finished",
      },
    },
  ],
}
`;
